File: frontend/src/api/addressbooks.ts
import apiClient from "./client";
import type {
  AddressBook,
  Contact,
  PaginatedResponse,
  DetectedColumn,
  ColumnMapping,
} from "@/types";

export async function listAddressBooks(params?: {
  page?: number;
  page_size?: number;
}): Promise<PaginatedResponse<AddressBook>> {
  const response = await apiClient.get<PaginatedResponse<AddressBook>>(
    "/address-books",
    { params }
  );
  return response.data;
}

export async function getAddressBook(id: string): Promise<AddressBook> {
  const response = await apiClient.get<AddressBook>(`/address-books/${id}`);
  return response.data;
}

export async function uploadAddressBook(
  file: File,
  name: string,
  description?: string
): Promise<AddressBook> {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("name", name);
  if (description) formData.append("description", description);
  const response = await apiClient.post<AddressBook>(
    "/address-books/upload",
    formData,
    { headers: { "Content-Type": "multipart/form-data" } }
  );
  return response.data;
}

export async function getContacts(
  addressBookId: string,
  params?: { page?: number; page_size?: number; search?: string }
): Promise<PaginatedResponse<Contact>> {
  const response = await apiClient.get<PaginatedResponse<Contact>>(
    `/address-books/${addressBookId}/contacts`,
    { params }
  );
  return response.data;
}

export async function detectColumns(
  addressBookId: string
): Promise<DetectedColumn[]> {
  const response = await apiClient.get<DetectedColumn[]>(
    `/address-books/${addressBookId}/detect-columns`
  );
  return response.data;
}

export async function mapColumns(
  addressBookId: string,
  mappings: ColumnMapping[]
): Promise<AddressBook> {
  const response = await apiClient.post<AddressBook>(
    `/address-books/${addressBookId}/map-columns`,
    { mappings }
  );
  return response.data;
}

export async function deleteAddressBook(id: string): Promise<void> {
  await apiClient.delete(`/address-books/${id}`);
}

File: backend/app/api/addressbooks.py
"""Address book and contact management router."""

from fastapi import APIRouter, Depends, UploadFile, File
from pydantic import BaseModel

from app.auth.dependencies import get_current_user
from app.schemas.common import PaginatedResponse, SuccessResponse

router = APIRouter(prefix="/addressbooks")


# -- Local schemas ---------------------------------------------------------


class AddressBookResponse(BaseModel):
    """Address book summary returned by the API."""

    id: int
    name: str
    contact_count: int
    created_at: str


class ContactResponse(BaseModel):
    """Single contact record."""

    id: int
    email: str
    first_name: str | None = None
    last_name: str | None = None
    position: str | None = None


class ColumnMapping(BaseModel):
    """Column mapping payload for CSV/XLSX imports."""

    email: str
    first_name: str | None = None
    last_name: str | None = None
    position: str | None = None


# -- Endpoints -------------------------------------------------------------


@router.post(
    "/addressbooks/upload",
    response_model=SuccessResponse,
    status_code=201,
)
async def upload_addressbook(
    file: UploadFile = File(...),
    _user: dict = Depends(get_current_user),
) -> dict:
    """Upload a CSV or XLSX file to create a new address book."""
    return {
        "message": f"File '{file.filename}' uploaded. Map columns to continue.",
        "data": {"addressbook_id": 1, "detected_columns": []},
    }


@router.get("/addressbooks", response_model=list[AddressBookResponse])
async def list_addressbooks(
    _user: dict = Depends(get_current_user),
) -> list[dict]:
    """Return all address books for the current user."""
    return []


@router.get("/addressbooks/{book_id}", response_model=AddressBookResponse)
async def get_addressbook(
    book_id: int,
    _user: dict = Depends(get_current_user),
) -> dict:
    """Return address book details including import statistics."""
    return {
        "id": book_id,
        "name": "Placeholder",
        "contact_count": 0,
        "created_at": "2026-01-01T00:00:00Z",
    }


@router.get(
    "/addressbooks/{book_id}/contacts",
    response_model=PaginatedResponse[ContactResponse],
)
async def list_contacts(
    book_id: int,
    page: int = 1,
    per_page: int = 50,
    _user: dict = Depends(get_current_user),
) -> dict:
    """Return paginated contacts for an address book."""
    return {
        "items": [],
        "total": 0,
        "page": page,
        "per_page": per_page,
        "pages": 0,
    }


@router.post(
    "/addressbooks/{book_id}/map-columns",
    response_model=SuccessResponse,
)
async def map_columns(
    book_id: int,
    mapping: ColumnMapping,
    _user: dict = Depends(get_current_user),
) -> dict:
    """Apply column mapping to a previously uploaded file and import contacts."""
    return {"message": f"Column mapping applied to address book {book_id}."}
